dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/requests/McpToolCallParams.java

```diff
@@ -71,32 +82,68 @@ public Map<String, Object> getArguments(Jsonb jsonb) {
-                result.put(argName, jsonb.fromJson(json, argMetadata.type()));
+            if (requestArguments.containsKey(argName)) {
+                JsonValue argValue = requestArguments.get(argName);
+                String argValueJson = jsonb.toJson(argValue);
+                result.put(argName, jsonb.fromJson(argValueJson, argMetadata.type()));
+            } else if (!argMetadata.required()) {
+                String defaultValue = argMetadata.defaultValue();
+                if (!defaultValue.isEmpty()) {
+                    result.put(argName, getDefaultValue(argMetadata));
+                } else {
+                    result.put(argName, emptyToolArgValue((Class<?>) argMetadata.type())); //blank result for no value provided for optional argument
+                }
+            } else {
+                List<String> data = generateArgumentMismatchData(requestArguments.keySet(), metadatas.keySet());
+                throw new JSONRPCException(JSONRPCErrorCode.INVALID_PARAMS, data);
             }
-            argsProcessed.add(argName);
-        }
-
-        validateProcessedArgs(argsProcessed, metadatas.keySet());
 
+        }
         return result;
     }
 
-    private void validateProcessedArgs(Set<String> processedArgs, Set<String> expectedArgs) {
-        if (!processedArgs.equals(expectedArgs)) {
-            List<String> data = generateArgumentMismatchData(processedArgs, expectedArgs);
-            throw new JSONRPCException(JSONRPCErrorCode.INVALID_PARAMS, data);
+    /**
+     * The null value to use for different types. Null for objects or 0 for primitives.
+     *
+     * @param type the type to get the null value for
+     * @return the null value for the class inputted as a parameter
+     */
+    public static Object emptyToolArgValue(Class<?> type) {
+        if (!type.isPrimitive())
+            return null;
+        return TYPE_DEFAULTS_MAP.get(type);
+    }
+
+    /**
+     * Converts a the value stored in a string to a Java object of type {@code type}
+     *
+     * @param value the string to be converted into a Java object
+     * @param type the type of Java object you want the string to be converted to
+     * @return the value as a Java object of type {@code type}
+     */
+    public static Object getDefaultValue(ArgumentMetadata argMetadata) {
+        String defaultValue = argMetadata.defaultValue();
+        Type type = TypeUtility.box(argMetadata.type());
+        DefaultValueConverter<?> converter = BuiltinDefaultValueConverters.CONVERTERS.get(type);
+        if (converter != null) {
+            return converter.convert(defaultValue);
+        }
+        if (type instanceof Class clazz) {
+            if (clazz.isEnum()) {
+                for (Object constant : clazz.getEnumConstants()) {
+                    if (constant.toString().equalsIgnoreCase(defaultValue)) {
+                        return constant;
+                    }
+                }
```

Comment:
```suggestion
                return Enum.valueOf(clazz.asSubclass(Enum.class), defaultValue);
```

This loses case-insensitivity, but I think that's fine for something the developer is providing as a default.

Reply:
It also occurs to be that we should probably parse out the default value at startup, and throw an exception then if there's an error.